File: src/main/scala/bank/Account.scala
package bank

import akka.actor.{AbstractLoggingActor, Props}
import bank.Account._

/**
 * Created by Rahul Shukla <rahul.shukla@synerzip.com> on 8/10/15.
 */

/**
 * API for Account actor: Messages, exceptions etc.
 */
object Account {



  /** Base message type. */
  sealed trait AccountMessage

  case object Balance extends AccountMessage

  /** Message to disposit the given amount. */
  case class Deposit(amount: Int) extends AccountMessage

  /** Message to withdraw the given amount. */
  case class Withdraw(amount: Int) extends AccountMessage

  case class Transfer( to: String,amount: Int) extends AccountMessage

  /** Aborts a withdrawl. */
  class OverdrawException extends RuntimeException

  def props(name: String): Props = {
    return Props(Account(name))
  }
}

case class Account(name: String) extends AbstractLoggingActor {
  var balance = 0
  log.info(s"Account Created $name with path ${self.path}" )

  def deposit(amount: Int) = {
    balance = balance + amount
    log.info(s"Deposit Success for Account  : $name . Current balance is => $balance")
  }

  def withdraw(amount: Int) = {
    if (amount > balance) {
      log.warning("Account overdrawn! Aborting by throwing an OverdrawException.")
      throw new OverdrawException
    } else {
      balance = balance - amount
      log.info(s"Withdraw Success for Account  : $name . Current balance is => $balance")
    }
  }

  def transfer( to: String,amount: Int)={
    if (amount <= balance) {
      withdraw(amount)
      sender ! Bank.Deposit(to,amount)
    }else{
      log.warning("Account overdrawn! Aborting by throwing an OverdrawException.")
      throw new OverdrawException
    }

  }

  override def receive = {
    case Balance => log.info(s"Balance for Account : $name is => $balance")
    case Deposit(amount) => deposit(amount)
    case Withdraw(amount) => withdraw(amount)
    case Transfer(to,amount) => transfer(to,amount)
  }

}

File: src/main/java/bank1/Account.java
package bank1;

import akka.actor.AbstractLoggingActor;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

/**
 * Created by Rahul Shukla <rahul.shukla@synerzip.com> on 9/10/15.
 */
public class Account extends AbstractLoggingActor {

    private final String name;
    private int balance = 0;

    public Account(String name) {
        this.name = name;
        log().info("Account Created " + name + "with path " + self().path());
    }

    /**
     * Base message type.
     */
    static interface AccountMessage {
    }

    static class Balance implements AccountMessage {
    }

    /**
     * Message to disposit the given amount.
     */
    static class Deposit implements AccountMessage {
        public final int amount;

        public Deposit(int amount) {
            this.amount = amount;
        }
    }

    /**
     * Message to withdraw the given amount.
     */
    static class Withdraw implements AccountMessage {
        public final int amount;

        public Withdraw(int amount) {

            this.amount = amount;
        }
    }

    static class Transfer implements AccountMessage {
        public final String to;
        public final int amount;

        public Transfer(String to, int amount) {
            this.to = to;
            this.amount = amount;
        }
    }

    /**
     * Aborts a withdrawl.
     */
    static class OverdrawException extends RuntimeException {
    }

    public static Props props(String name) {
        return Props.create(Account.class, () -> new Account(name));
    }


    private void deposit(int amount) {
        balance = balance + amount;
        log().info("Deposit Success for Account  : " + name + ". Current balance is => " + balance);
    }

    private void withdraw(int amount) {
        if (amount > balance) {
            log().warning("Account overdrawn! Aborting by throwing an OverdrawException.");
            throw new OverdrawException();
        } else {
            balance = balance - amount;
            log().info("Withdraw Success for Account  : " + name + ". Current balance is => " + balance);
        }
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receive() {
        return ReceiveBuilder
                .match(Balance.class, b -> log().info("Balance for Account :" + name + " is =>" + balance))
                .match(Deposit.class, d -> deposit(d.amount))
                .match(Withdraw.class, w -> withdraw(w.amount))
                .match(Transfer.class, t -> transfer(t.to, t.amount))
                .build();
    }

    private void transfer(String to, int amount) {
        if(balance >= amount){
            withdraw(amount);
            sender().tell(new Bank.Deposit(to,amount),self());
        }
    }


}
